### ISSUE TITLE
2019-11-29 tc008 M4SD crash

### ISSUE THREAD

**Issue by tensfeldt** [tensfeldt/openNCA#67]
With commit 22d43ea tc008 M4 SD doesn't execute

Note that while AURCT1_T2 is requested with PARAMETERLIST, there are no defined partial AUCs. This should not cause a failure. With this situation, the computation engine should generate no partial AUCs but not cause a crash.

Regardless this, when AURCT1_T2 is removed from mct$PARAMETERLIST, the same crash occurs (see below).

Note that there is only one sample collection interval / profile but this test cases was running prior to this commit.

```
> d[,c("SDEID", "SUBJID", "PKPTMS", "PKATPD", mct$NOMENDTIME, mct$ACTENDTIME)]
      SDEID   SUBJID PKPTMS PKATPD PKPTME PKATPDE
1 363476907 10011001      0      0     24  24.083
2 754426795 10011002      0      0     24  24.083
3 926177600 10011004      0      0     24  24.083
4 927450954 10011005      0      0     24  24.083
5 225930823 10011007      0      0     24  24.133
6 532644281 10011008      0      0     24  24.083
```


```
> mct[, names(parameter_required("^AUC.[0-9]+?.T[1-2]$", names(mct), simplify=FALSE))]
data frame with 0 columns and 1 row
```

## mct

```
 MODEL DOSINGTYPE AUCMETHOD SDEID NOMTIME NOMTIMEU ACTTIME ACTTIMEU   CONC  CONCU AMOUNT AMOUNTU NOMENDTIME ACTENDTIME FLGEXKEL FLGEXAUC FLGEXSDE
1    M4         SD    LINLOG SDEID  PKPTMS   PKPTMU  PKATPD  PKATPDU PKCNCN PKCNCU PKSMVL PKSMVLU     PKPTME    PKATPDE FLGEXKEL FLGEXAUC FLGEXSDE
  FLGEXST  MATRIX PKPCOM TIMEOUTPUTUNIT AMOUNTOUTPUTUNIT DOSEOUTPUTUNIT VOLUMEOUTPUTUNIT CONCOUTPUTUNIT KELOUTPUTUNIT CLOUTPUTUNIT AUCOUTPUTUNIT
1 FLGEXST PKBDFLD PKPCOM             HR               MG             MG                L          NG/ML          1/HR         L/HR      NG.HR/ML
  AUMCOUTPUTUNIT AUCNORMOUTPUTUNIT AUROUTPUTUNIT CONCNORMOUTPUTUNIT AUCNPAIR DP.AE DP.AEPCT DP.AET DP.AETPCT DP.CLAST DP.KEL DP.KELRSQ DP.KELRSQA
1    NG.HR.HR/ML       NG.HR/ML/GM      ML.UG/ML           NG/ML/MG        0     3        3      3         3        3      3         4          4
  DP.KELTMHI DP.KELTMLO DP.THALF DP.TLAG DP.TLAST SF.AE SF.AEPCT SF.AET SF.AETPCT SF.CLAST SF.KEL SF.KELRSQ SF.KELRSQA SF.KELTMHI SF.KELTMLO SF.THALF
1          3          3        3       3        3     3        3      3         3        3      3         4          4          3          3        3
  SF.TLAG SF.TLAST SAMPLETYPE
1       3        3     PKCOLL
                                                                                                                                                                                                                                                                                                                                                                                                                                         RETURNCOLS
1 UDSDEID;SDEID;STUDY;SITEID;SUBJID;RAND;TREATXT;TRTCD;PKCOLL;PKBDFLD;PKTERM;PERIODU;PERIOD;VISITU;VISIT;PHASE;PCANMETH;COLLDATE;HT;WT;AGEDERU;AGEDER;WTUNI;WTRAW;HTUNI;HTRAW;RACEOTH;RACES;SEX;RACIALD;ETHNIC;DOSEP;DOSPND;DOSEPUNI;DOSPTM;DOSE;DOSEUNI;ACTTRT;ACTTRTC;ACTTRTS;TREATSEQ;FLGEMESIS;FU;UDFNAME1;UDFVALUE1;UDFUNIT1;UDFNAMEn;UDFVALUEn;UDFUNITn;DATASTATUS;PKPCOM;PKRCOM;PKSIGFIG;PKDECPL;DATABLINDSTATUS;PCLLOQ;PCSTRESU;PCNAM;ROUTE
  PKADT PKATM PKADTE PKATME DRGDATE  DOSETIM DRGDATESTOP DOSETIMESTOP NOMENDTIMEU ACTENDTIMEU CONCRAW     LLOQPATTERNS SAMPLEVOLUME SAMPLEVOLUMEU
1 PKADT PKATM PKADTE PKATME DRGDATE DOSETIME DRGDATESTOP DOSETIMESTOP      PKPTMU    PKATPDEU   PKCNC LLOQ|BLQ|^[<]{1}       PKSMVL       PKSMVLU
  DOFI    FLGMERGE       FLGACCEPTKELCRIT FLGNOCMAX FLGACCEPTTAU FLGACCEPTPREDOSECRIT LASTTIMEACCEPTCRIT SPANRATIOCRIT NORMBS PKTERMPARENT
1  DOF PKDATAROWID KELRSQ>=0.9,KELNOPT>=3         1 FLGACCEPTTAU                 0.05         0.9 * TAUi             2     WT           NA
  RATEOUTPUTUNIT VOLUMENORMOUTPUTUNIT CLNORMOUTPUTUNIT
1          UG/HR                ML/KG         ML/HR/KG
                                                                                                                                                                                                                                                                                                         PARAMETERLIST
1 AE;AEPCT;AET;AETPCT;AT;AET;AETAU;AETAUi;AURCALL;AURCINFO;AURCINFP;AURCLAST;AURCT;AURCT1_T2;AURCXPCTO;AURCXPCTP;DOSE1;KEL;KELNOPT;KELR;KELRSQ;KELRSQA;KELTMHI;KELTMLO;MAXRATE;MIDPTLAST;MIDPTA;MIDPTN;RATEA;RATELAST;RATEN;THALF;THALFF;TLAG;TMAXRATE;VOLSUM;FLGACCEPTKEL;FLGACCEPTPREDOSE;FLGACCEPTTAU;FLGACCEPTTMAX
                                                                                           PARAMETERDISPLAYLIST
1 AE;AEPCT;AURCALL;AURCINFP;AURCLAST;AURCXPCTP;DOSE1;KEL;KELRSQ;KELNOPT;MAXRATE;MIDPTLAST;THALF;THALFF;TMAXRATE
                                                                                                                  DATADISPLAYLIST DP.DEFAULT
1 SDEID;STUDY;SITEID;SUBJID;RAND;TREATXT;TRTCD;PKCOLL;PKBDFLD;PKTERM;PERIODU;PERIOD;VISITU;VISIT;PHASE;PCANMETH;DOSE;DOSEU;ACTTRT          3
  SF.DEFAULT DOSE1   DOSEU    TIME OPTIMIZEKEL INCLUDEINTERPOLATION INCLUDEEXTRAPOLATION TAU1 TAU1U TOLD1 TOLD1U
1          3  DOSE DOSEUNI NOMINAL           0                    1                    1  TAU TAU1U  TOLD TOLD1U
```

## execution fails

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTPREDOSECRIT' is not present in the dataset
3: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823 
```
## remove AURCT1_T2 from PARAMETERLIST

```
> mct$PARAMETERLIST <- gsub('AURCT1_T2;', '', mct$PARAMETERLIST, ignore.case=TRUE, perl=TRUE)
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTPREDOSECRIT' is not present in the dataset
3: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823 
```

**Comment by rudraya** [comment]
I reran this using the update MCT that was provided, but the computation executed without any errors. Not sure what is causing this error that is mentioned above.

**Comment by tensfeldt** [comment]
This error still exists with commit 6e99fab

**Comment by rudraya** [comment]
I tested again using the script you provided (with minor modifications) and I was able to run fine. I did notice that you were reading the data from the JSON files rather than CSV file. Not sure if the difference is in that, but I am not able to recreate this method. I am able to run tc008 using the CSV files that are uploaded as of the latest commit.

**Comment by tensfeldt** [comment]
I can still reproduce this with commit bb8e82d. The error message implicates the auc_t1_t2 function and since it occurs whether partial AURCs are included or not, it's likely coming not from the definition of the partial AURCs. This function is called in 3 different places in run_M4_SD_computation.R. What were the minor modifications that you implemented prior to running with the MCT (in CSV form - I only re-read the JSON file since that's the standard workflow in the openNCA application)? Can you attempt using the JSON file directly for MCT rather than CSV in case there's something with that causing this.



```
Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823 
```

**Comment by tensfeldt** [comment]
If it helps, this is occurring in the AURCT call to auc_t1_t2: (I've added some debugging code to run_M4_SD_computation.R to determine what it's complaining about and it appears to be related directly to computing the incremental AURCTs

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
t:  2  mid_pt:  12  mid_pt[1]:  12  mid_pt[t]:  NA 
t:  1  mid_pt:  12  mid_pt[1]:  12  mid_pt[t]:  12 
Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823
```
If AURCT is removed from PARAMETERLIST, the the following different error is obtained:

```
> mct$PARAMETERLIST
[1] "AE;AEPCT;AET;AETPCT;AT;AET;AETAU;AETAUi;AURCALL;AURCINFO;AURCINFP;AURCLAST;AURCT1_T2;AURCXPCTO;AURCXPCTP;DOSE1;KEL;KELNOPT;KELR;KELRSQ;KELRSQA;KELTMHI;KELTMLO;MAXRATE;MIDPTLAST;MIDPTA;MIDPTN;RATEA;RATELAST;RATEN;THALF;THALFF;TLAG;TMAXRATE;VOLSUM;FLGACCEPTKEL;FLGACCEPTPREDOSE;FLGACCEPTTAU;FLGACCEPTTMAX;FLGACCEPTKELCRIT;FLGACCEPTPREDOSECRIT;SPANRATIOCRIT"
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (is.null(conc1) || is.na(conc1)) {: missing value where TRUE/FALSE needed
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (is.null(conc1) || is.na(conc1)) {: missing value where TRUE/FALSE needed
For SDEID 225930823 
```

if AURCT is also removed from PARAMETERLIST then the computation call completes successfully but without the incremental AURCT value.

Recall for this particular testcase tc008 M4SD, there is only the following sampling data
```
> table(d$PKPTMS)
0 
6 
> table(d$PKPTME)
24 
 6  
```

```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
unit_conversion : Unit Class 1 (Time) time_col:  TMAXRATE MIDPTLAST TLAG KELTMLO KELTMHI THALF MIDPT1  parameters are scaled from  HR  to  HR  via scaling factor:  1 
unit_conversion : Unit Class 2 (Amount) amount_col:  AMT.24.00 AE.24.00 AE  parameters are scaled from  NG  to  MG  via scaling factor:  1e-06 
unit_conversion : Unit Class 3 (Dose) dose_col:  DOSE  parameters are scaled from  MG  to  MG  via scaling factor:  1 
unit_conversion : Unit Class 4 (Volume) volume_col:  VOLSUM  parameters are scaled from  ML  to  L  via scaling factor:  0.001 
unit_conversion : Unit Class 6 (1/Time) kel_col:  KEL  parameters are scaled from  1/HR  to  1/HR  via scaling factor:  1 
unit_conversion : Unit Class 11: ([Volume.Amount]/Volume) aurc_col:  AURCALL AURCLAST AURCINFO AURCINFP  parameters are scaled from  ML.NG/ML  to  ML.UG/ML  via scaling factor:  0.001 
unit_conversion : Unit Class 13: (Amount/Time) rate_col:  MAXRATE RATELAST RATE1  parameters are scaled from  NG/HR  to  UG/HR  via scaling factor:  0.001 
Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTPREDOSECRIT' cannot be computed if 'CMAX' is not part of the calculated parameters
3: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'PKPTMU', 'CONCU' and 'DOSEU' #8# are not present in the dataset provided via 'map'
4: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU' #24# is not present in the dataset provided via 'map'
5: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU' and 'NORMBSU' #14# are not present in the dataset provided via 'map'
6: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'CONCU', 'PKPTMU and 'NORMBSU' #19# are not present in the dataset provided via 'map'
```

**Comment by tensfeldt** [comment]
Note that if only AURCT is removed from PARAMETERLIST, leaving AURCT1_T2 in PARAMETERLIST then the following error is obtained:

```
> mct$PARAMETERLIST
[1] "AE;AEPCT;AET;AETPCT;AT;AET;AETAU;AETAUi;AURCALL;AURCINFO;AURCINFP;AURCLAST;AURCT1_T2;AURCXPCTO;AURCXPCTP;DOSE1;KEL;KELNOPT;KELR;KELRSQ;KELRSQA;KELTMHI;KELTMLO;MAXRATE;MIDPTLAST;MIDPTA;MIDPTN;RATEA;RATELAST;RATEN;THALF;THALFF;TLAG;TMAXRATE;VOLSUM;FLGACCEPTKEL;FLGACCEPTPREDOSE;FLGACCEPTTAU;FLGACCEPTTMAX;FLGACCEPTKELCRIT;FLGACCEPTPREDOSECRIT;SPANRATIOCRIT"
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in if (is.null(conc1) || is.na(conc1)) {: missing value where TRUE/FALSE needed
For SDEID 225930823
In addition: Warning messages:
1: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGEMESIS' is not present in the dataset
2: In run_M4_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in if (is.null(conc1) || is.na(conc1)) {: missing value where TRUE/FALSE needed
For SDEID 225930823 
```

**Comment by tensfeldt** [comment]
Note that this issue occurs when ACTUAL time (mct$TIME == "actual") regardless whether AURCT nor AURCT1_T2 is specified in the MCT$PARAMETERLIST so it's a bit more involved than simply excluding AURCT and AURCT1_T2 from the parameterlist.

**Comment by jhhughes256** [comment]
### Update

I am able to replicate this error with tc008 M4SDSS after using the dataset fix mentioned in #77. I get the following error for the base test case and when removing all partial AURCs (`AUC.1.T1 <- NULL` etc.).

```
 Error in value[[3L]](cond) : 
  Error in auc_t1_t2(conc = rt, time = mid_pt, t1 = mid_pt[1], t2 = mid_pt[t], : Error in auc_t1_t2: 't1' value is greater than 't2' value
For SDEID 225930823 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M4_SD_computation.R#1514
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    tmp_df <- data_data[data_data[, map_data$SDEID] == unique(data_data[, 
        map_data$SDEID])[i], ]
    tmp_df <- tmp_df[order(tmp_df[, map_data$TIME]), ] ... at run_M4_SD_computation.R#742
2. run_M4_SD_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2326
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset) 
```

**Comment by rudraya** [comment]
This issue was partially addressed in f98f923, please let me know if it still occurs. I will take a more in-depth look if needed once I acquire the proper test data.

**Comment by rudraya** [comment]
This issue was addressed in fce7a3c, I checked tc008 M1SD, M1SS and M4SDSS, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
### Update

As of commit fce7a3c there are no computation errors for unchanged tc008 M1SD, M1SS and M4SD. The following error occurs for tc008 M4SS (`mct$DOSINGTYPE <- "SS"`). This is believed to be due to "AUCRT" as stated in #86 .

```
 Error in value[[3L]](cond) : 
  Error in aurct[[t - 1]] <- tmp: attempt to select less than one element in integerOneIndex
For SDEID 225930823 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M4_SS_computation.R#1513
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    if (comp_required[["DIi"]]) {
        di <- list()
    } ... at run_M4_SS_computation.R#688
2. run_M4_SS_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2389
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset)
```

**Comment by rudraya** [comment]
This issue was addressed in bad1e92, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
The error for M4SS no longer occurs as of b1c7c7072a.
